src/main/java/it/smartcommunitylab/aac/spid/auth/SpidWebSsoAuthenticationRequestFilter.java

```diff
@@ -0,0 +1,348 @@
+import org.springframework.security.saml2.provider.service.web.Saml2AuthenticationRequestContextResolver;
+import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+import org.springframework.web.util.HtmlUtils;
+import org.springframework.web.util.UriComponentsBuilder;
+import org.springframework.web.util.UriUtils;
+
+/*
+ * SpidWebSsoAuthenticationRequestFilter is the filter that intercepts SPID authentication requests, generates the SAML
+ * request and then sends the sso authentication request to the selected (SPID) SAML identity provider.
+ * This is a filter mostly aligned with the with the Spring Security architecture.
+ * For more on how SPID authentication requests are made, see
+ *  https://docs.italia.it/italia/spid/spid-regole-tecniche/it/stabile/single-sign-on.html
+ * For more on the Spring Security architecture and why an AuthenticationRequestFilter is required, see
+ *  https://docs.spring.io/spring-security/reference/servlet/saml2/login/overview.html
+ */
+public class SpidWebSsoAuthenticationRequestFilter
+    extends OncePerRequestFilter
+    implements ApplicationEventPublisherAware {
+
+    private static final String DEFAULT_FILTER_URI =
+        SpidIdentityAuthority.AUTHORITY_URL + "authenticate/{registrationId}"; // NOTE: if multiple upstream idps are possible, then registration id must somehow also encode which idp is target of the authentication request
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final String authorityId;
+    private final RequestMatcher requestMatcher;
+
+    private final Saml2AuthenticationRequestContextResolver authenticationRequestContextResolver;
+    private final Saml2AuthenticationRequestFactory authenticationRequestFactory;
+    private ApplicationEventPublisher eventPublisher;
+    private final ProviderConfigRepository<SpidIdentityProviderConfig> providerConfigRepository;
+
+    private Saml2AuthenticationRequestRepository<SerializableSaml2AuthenticationRequestContext> authenticationRequestRepository =
+        new HttpSessionSaml2AuthenticationRequestRepository(
+            HttpSessionSaml2AuthenticationRequestRepository.class.getName() + ".SPID_AUTHORIZATION_REQUEST"
+        );
+
+    public SpidWebSsoAuthenticationRequestFilter(
+        ProviderConfigRepository<SpidIdentityProviderConfig> providerConfigRepository,
+        RelyingPartyRegistrationRepository relyingPartyRegistrationRepository
+    ) {
+        this(
+            SystemKeys.AUTHORITY_SPID,
+            providerConfigRepository,
+            relyingPartyRegistrationRepository,
+            DEFAULT_FILTER_URI
+        );
+    }
+
+    public SpidWebSsoAuthenticationRequestFilter(
+        String authorityId,
```

Comment:
you can drop the customizable authorityId from every class, this is SPID specific we do *want* to avoid reuse